osu.Game.Rulesets.Taiko/Difficulty/Preprocessing/Colour/Data/CoupledColourEncoding.cs

```diff
@@ -0,0 +1,89 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Collections.Generic;
+
+namespace osu.Game.Rulesets.Taiko.Difficulty.Preprocessing.Colour.Data
+{
+    /// <summary>
+    /// Encodes a list of <see cref="ColourEncoding"/>s, grouped together by back and forth repetition of the same
+    /// <see cref="ColourEncoding"/>. Also stores the repetition interval between this and the previous <see cref="CoupledColourEncoding"/>.
+    /// </summary>
+    public class CoupledColourEncoding
+    {
+        /// <summary>
+        /// Maximum amount of <see cref="CoupledColourEncoding"/>s to look back to find a repetition.
+        /// </summary>
+        private const int max_repetition_interval = 16;
+
+        /// <summary>
+        /// The <see cref="ColourEncoding"/>s that are grouped together within this <see cref="CoupledColourEncoding"/>.
+        /// </summary>
+        public readonly List<ColourEncoding> Payload = new List<ColourEncoding>();
+
+        /// <summary>
+        /// The previous <see cref="CoupledColourEncoding"/>. This is used to determine the repetition interval.
+        /// </summary>
+        public readonly CoupledColourEncoding? Previous;
+
+        /// <summary>
+        /// How many <see cref="CoupledColourEncoding"/> between the current and previous identical <see cref="CoupledColourEncoding"/>.
+        /// If no repetition is found this will have a value of <see cref="max_repetition_interval"/> + 1.
+        /// </summary>
+        public int RepetitionInterval { get; private set; } = max_repetition_interval + 1;
+
+        public CoupledColourEncoding(CoupledColourEncoding? previous)
+        {
+            Previous = previous;
+        }
+
+        /// <summary>
+        /// Returns true if other is considered a repetition of this encoding. This is true if other's first two payloads
+        /// have identical mono lengths.
+        /// </summary>
+        private bool isRepetitionOf(CoupledColourEncoding other)
+        {
+            if (Payload.Count != other.Payload.Count) return false;
+
+            for (int i = 0; i < Math.Min(Payload.Count, 2); i++)
```

Comment:
Why does this only check the first two payloads?

As far as I understand, each "payload" here is a sequence of same-length colour-alternating mono patterns. But that doesn't mean that each pattern in the "payload" has the same length of monos? What am I missing?

Reply:
This is intended - as we were tweaking this we wanted `findRepetitionInterval` to have a more lenient definition of repetition (side note, it was actually a strict comparison originally), so that we can get repetition interval to catch more patterns. For example, we want to consider `ddkkddkkdkdkdkdk` as a repetition of `ddkkdkdkddkkdkdk`

Here it is defined as "having the same mono length (how many notes within a single `MonoEncoding`) encoded within it's children `ColourEncoding`s". Because CoupledColourEncoding encodes ColourEncoding of back-and-forth repetitions of the same mono length, the subsequent ColourEncodings after the first two are guaranteed to have the same repeating pattern of mono lengths.

Reply:
Right, I see now that the even and odd alternating mono patterns in a `CoupledColourEncoding` are guaranteed to have equal mono lengths, as it's enforced by the final encoding step.